server/app/controllers/applicant/ApplicantProgramBlocksController.java

```diff
@@ -134,6 +146,51 @@ public CompletionStage<Result> review(
     return editOrReview(request, applicantId, programId, blockId, true);
   }
 
+  /** This method handles the applicant's selection from the address correction options. */
+  @Secure
+  public CompletionStage<Result> confirmAddress(
+      Request request, long applicantId, long programId, String blockId, boolean inReview) {
+
+    DynamicForm form = formFactory.form().bindFromRequest(request);
+    String selectedAddress = form.get(AddressCorrectionBlockView.SELECTED_ADDRESS_NAME);
+    Optional<String> maybeAddressJson = request.session().get(ADDRESS_JSON_SESSION_KEY);
+
+    ImmutableList<AddressSuggestion> suggestions =
+        addressSuggestionJsonSerializer.deserialize(
+            maybeAddressJson.orElseThrow(() -> new RuntimeException("Address JSON missing")));
+
+    return checkApplicantAuthorization(profileUtils, request, applicantId)
+        .thenComposeAsync(
+            v -> applicantService.getReadOnlyApplicantProgramService(applicantId, programId),
+            httpExecutionContext.current())
+        .thenComposeAsync(
+            roApplicantProgramService ->
+                applicantService.getCorrectedAddress(
+                    applicantId, programId, blockId, selectedAddress, suggestions),
+            httpExecutionContext.current())
+        .thenComposeAsync(
+            questionPathToValueMap ->
+                applicantService.stageAndUpdateIfValid(
+                    applicantId,
+                    programId,
+                    blockId,
+                    cleanForm(questionPathToValueMap),
+                    featureFlags.isEsriAddressServiceAreaValidationEnabled(request)),
+            httpExecutionContext.current())
+        .thenComposeAsync(
+            roApplicantProgramService ->
+                renderErrorOrRedirectToNextBlock(
+                    request,
+                    applicantId,
+                    programId,
+                    blockId,
+                    applicantService.getName(applicantId).toCompletableFuture().join(),
```

Comment:
**issue (blocking):** `join` here is blocking the HTTP thread on I/O

Reply:
@bion   I'm not sure how to change this. This is what the other endpoints in this class are doing. Can you point me to an example of what you'd like to see?

Reply:
It's acceptable to do if we know the future has already completed. In the other methods of this class we call `join()` in a callback that's triggered (at some point) after the completion of the future.

Here's a diff that resolves the issue:
```
--- a/server/app/controllers/applicant/ApplicantProgramBlocksController.java
+++ b/server/app/controllers/applicant/ApplicantProgramBlocksController.java
@@ -158,11 +158,11 @@ public final class ApplicantProgramBlocksController extends CiviFormController {
     ImmutableList<AddressSuggestion> suggestions =
         addressSuggestionJsonSerializer.deserialize(
             maybeAddressJson.orElseThrow(() -> new RuntimeException("Address JSON missing")));
+    CompletableFuture<Optional<String>> applicantNameStage =
+        applicantService.getName(applicantId).toCompletableFuture();
 
-    return checkApplicantAuthorization(profileUtils, request, applicantId)
-        .thenComposeAsync(
-            v -> applicantService.getReadOnlyApplicantProgramService(applicantId, programId),
-            httpExecutionContext.current())
+    return CompletableFuture.allOf(
+            checkApplicantAuthorization(profileUtils, request, applicantId), applicantNameStage)
         .thenComposeAsync(
             roApplicantProgramService ->
                 applicantService.getCorrectedAddress(
@@ -184,7 +184,7 @@ public final class ApplicantProgramBlocksController extends CiviFormController {
                     applicantId,
                     programId,
                     blockId,
-                    applicantService.getName(applicantId).toCompletableFuture().join(),
+                    applicantNameStage.join(),
                     inReview,
                     roApplicantProgramService),
             httpExecutionContext.current())
```

Reply:
Oh, ok. I get it. Thank you! Done